File: packages/desktop-host/engine/src/protocol.rs
//! Wire types for the local control protocol and for the session's control
//! channel. Documented in `docs/PROTOCOL.md`; this module is the definition.

use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u32 = 1;

/// A request from the consumer to the engine.
#[derive(Debug, Deserialize)]
pub struct Request {
    #[serde(default)]
    pub id: Option<u64>,
    pub method: String,
    #[serde(default)]
    pub params: serde_json::Value,
}

/// A reply to one request. Exactly one of `result`/`error` is set.
#[derive(Debug, Serialize)]
pub struct Response {
    pub id: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<ErrorBody>,
}

#[derive(Debug, Serialize)]
pub struct ErrorBody {
    pub code: String,
    pub message: String,
}

impl ErrorBody {
    pub fn new(code: &str, message: impl Into<String>) -> Self {
        Self {
            code: code.to_owned(),
            message: message.into(),
        }
    }
}

/// A notification the engine sends on its own.
#[derive(Debug, Serialize)]
pub struct Event {
    pub event: String,
    pub params: serde_json::Value,
}

#[derive(Debug, Default, Deserialize)]
pub struct HelloParams {
    #[serde(default)]
    pub protocol: u32,
    #[serde(default)]
    pub client: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Permission {
    View,
    Control,
    Clipboard,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IceServerParam {
    #[serde(default)]
    pub urls: Vec<String>,
    #[serde(default)]
    pub username: Option<String>,
    #[serde(default)]
    pub credential: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct OpenParams {
    #[serde(default)]
    pub permissions: Vec<Permission>,
    #[serde(default = "default_max_width")]
    pub max_width: usize,
    #[serde(default = "default_max_height")]
    pub max_height: usize,
    #[serde(default = "default_bitrate")]
    pub bitrate_kbps: u32,
    #[serde(default = "default_fps")]
    pub max_fps: u32,
    #[serde(default)]
    pub ice_servers: Vec<IceServerParam>,
    #[serde(default)]
    pub relay_only: bool,
    #[serde(default)]
    pub restore_token: Option<String>,
    #[serde(default)]
    pub ttl_seconds: Option<u64>,
}

fn default_max_width() -> usize {
    1280
}
fn default_max_height() -> usize {
    800
}
fn default_bitrate() -> u32 {
    4000
}
fn default_fps() -> u32 {
    30
}

#[derive(Debug, Deserialize)]
pub struct SessionRef {
    pub session_id: String,
    #[serde(default)]
    pub generation: Option<u64>,
}

#[derive(Debug, Deserialize)]
pub struct DescriptionParams {
    pub session_id: String,
    #[serde(default)]
    pub generation: Option<u64>,
    pub description: RtcDescription,
}

#[derive(Debug, Deserialize)]
pub struct RtcDescription {
    #[serde(rename = "type")]
    pub kind: String,
    pub sdp: String,
}

#[derive(Debug, Deserialize)]
pub struct CandidateParams {
    pub session_id: String,
    #[serde(default)]
    pub generation: Option<u64>,
    pub candidate: String,
    #[serde(default)]
    pub sdp_mid: Option<String>,
    #[serde(default, rename = "sdpMLineIndex", alias = "sdp_m_line_index")]
    pub sdp_m_line_index: Option<u16>,
}

#[derive(Debug, Deserialize)]
pub struct ClipboardParams {
    pub session_id: String,
    #[serde(default)]
    pub mime: Option<String>,
    #[serde(default)]
    pub text: Option<String>,
}

/// Messages the client sends on the session's control channel.
///
/// `seq` is the client's own ordering. The engine refuses a repeat or a lower
/// value rather than re-applying an action, so a replayed frame cannot move the
/// pointer twice.
#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ControlMessage {
    Pointer {
        phase: PointerPhase,
        x: i64,
        y: i64,
        #[serde(default = "left_button")]
        button: i64,
        #[serde(default)]
        seq: u64,
    },
    Wheel {
        #[serde(default)]
        dx: i64,
        #[serde(default)]
        dy: i64,
        #[serde(default)]
        seq: u64,
    },
    Key {
        #[serde(default)]
        name: Option<String>,
        #[serde(default)]
        character: Option<String>,
        down: bool,
        #[serde(default)]
        modifiers: Vec<String>,
        #[serde(default)]
        seq: u64,
    },
    Text {
        text: String,
        #[serde(default)]
        seq: u64,
    },
    ReleaseAll {
        #[serde(default)]
        seq: u64,
    },
    ClipboardRead {
        request: String,
        #[serde(default)]
        seq: u64,
    },
    ClipboardWrite {
        request: String,
        text: String,
        #[serde(default)]
        seq: u64,
    },
}

fn left_button() -> i64 {
    1
}

impl ControlMessage {
    pub fn seq(&self) -> u64 {
        match self {
            Self::Pointer { seq, .. }
            | Self::Wheel { seq, .. }
            | Self::Key { seq, .. }
            | Self::Text { seq, .. }
            | Self::ReleaseAll { seq }
            | Self::ClipboardRead { seq, .. }
            | Self::ClipboardWrite { seq, .. } => *seq,
        }
    }
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PointerPhase {
    Move,
    Down,
    Up,
    Cancel,
}

/// Messages the engine sends back on the control channel.
#[derive(Debug, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ControlReply<'a> {
    Hello {
        protocol: u32,
        geometry: serde_json::Value,
    },
    Ack {
        seq: u64,
    },
    Rejected {
        seq: u64,
        code: &'a str,
        message: &'a str,
    },
    Clipboard {
        request: &'a str,
        text: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        error: Option<&'a str>,
    },
    Revoked {
        reason: &'a str,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_pointer_message_round_trips_with_its_sequence() {
        let parsed: ControlMessage =
            serde_json::from_str(r#"{"kind":"pointer","phase":"down","x":10,"y":20,"seq":3}"#)
                .unwrap();
        match parsed {
            ControlMessage::Pointer { phase, x, y, seq, button } => {
                assert!(matches!(phase, PointerPhase::Down));
                assert_eq!((x, y, seq, button), (10, 20, 3, 1));
            }
            other => panic!("wrong variant: {other:?}"),
        }
    }

    #[test]
    fn an_unknown_action_is_refused_rather_than_ignored() {
        assert!(serde_json::from_str::<ControlMessage>(r#"{"kind":"teleport"}"#).is_err());
    }

    #[test]
    fn open_defaults_are_a_phone_sized_surface() {
        let params: OpenParams = serde_json::from_str(r#"{"permissions":["view","control"]}"#).unwrap();
        assert_eq!(params.max_width, 1280);
        assert_eq!(params.max_height, 800);
        assert_eq!(params.max_fps, 30);
        assert!(!params.relay_only);
    }
}

File: packages/desktop-client/src/protocol.ts
/**
 * The wire contract this client speaks, in one place.
 *
 * Two layers, deliberately separate:
 *
 *  - **The engine's local control protocol** (lifecycle, SDP/ICE, clipboard,
 *    metrics). An application carries it over whatever authenticated channel it
 *    already has; this package never assumes a transport.
 *  - **The session's control channel** (pointer, wheel, keys, text), which rides
 *    the WebRTC data channel the engine creates. It belongs to the authorized
 *    media session, so revoking the session closes it.
 *
 * `docs/PROTOCOL.md` in the host engine package is the authority; these types
 * only describe the subset a client needs.
 */

export const PROTOCOL_VERSION = 2;

export type Permission = 'view' | 'control' | 'clipboard';

export interface IceServerConfig {
    urls: string[];
    username?: string;
    credential?: string;
}

/** Geometry of the surface, as the engine reports it after capture starts. */
export interface SurfaceGeometry {
    source: { width: number; height: number };
    encoded: { width: number; height: number };
    /** Origin of the captured source inside the desktop's own layout. */
    origin: { x: number; y: number };
}

export interface SessionOpenRequest {
    permissions: Permission[];
    maxWidth?: number;
    maxHeight?: number;
    bitrateKbps?: number;
    maxFps?: number;
    iceServers?: IceServerConfig[];
    restoreToken?: string;
    ttlSeconds?: number;
}

export interface SessionOpenResult {
    sessionId: string;
    generation: number;
    source: { kind: string; width: number; height: number; origin: { x: number; y: number } };
    geometry: SurfaceGeometry;
}

export interface RtcDescription {
    type: 'offer' | 'answer';
    sdp: string;
}

export interface RtcCandidate {
    candidate: string;
    sdpMid?: string | null;
    sdpMLineIndex?: number | null;
}

/** Engine → client lifecycle events, already unwrapped from the local protocol. */
export type SessionEvent =
    | { kind: 'description'; description: RtcDescription }
    | { kind: 'candidate'; candidate: RtcCandidate }
    | { kind: 'state'; capture: string; transport: string; firstFrame: boolean }
    | { kind: 'restoreToken'; token: string }
    | { kind: 'revoked'; reason: string };

/**
 * The application's own authenticated channel to the host engine.
 *
 * The package calls these and nothing else, so an application can carry them
 * over its existing encrypted request path, a socket, or an in-process bridge.
 * `subscribe` delivers engine notifications until the returned function runs.
 */
export interface Signaling {
    /**
     * Send one request. A rejection may carry a stable `code` — the engine's or
     * the host's own token — which the package reads to classify a refusal, so a
     * refusal is not retried as though it were a network failure.
     */
    request<T = unknown>(method: string, params?: Record<string, unknown>): Promise<T>;
    subscribe(handler: (event: SessionEvent) => void): () => void;
}

export type SessionStatus =
    | 'idle'
    | 'opening'
    | 'connecting'
    | 'live'
    | 'reconnecting'
    | 'ended'
    | 'failed';

export type SessionFailure =
    | { code: 'permission'; message: string }
    /** Nobody approved the desktop's own screen-sharing prompt in time. */
    | { code: 'consent'; message: string }
    /** The computer has no screen at all and cannot start one (a server without Xvfb). */
    | { code: 'no-screen'; message: string }
    | { code: 'unsupported-codec'; message: string }
    | { code: 'incompatible-version'; message: string }
    | { code: 'source-changed'; message: string }
    | { code: 'revoked'; message: string }
    | { code: 'transport'; message: string }
    | { code: 'input-unavailable'; message: string }
    | { code: 'platform'; message: string };

export interface SessionSnapshot {
    status: SessionStatus;
    /** Geometry is only known once the engine has a consented capture stream. */
    geometry: SurfaceGeometry | null;
    /** True once an authorized, current-generation frame has actually rendered. */
    presented: boolean;
    failure: SessionFailure | null;
    /** Redacted diagnostics: codec, route, counters. Never identities. */
    diagnostics: Record<string, string | number | boolean>;
}

/**
 * Control-channel messages. `seq` is monotonic per generation; the engine
 * refuses a repeat, so a replayed gesture cannot move the pointer twice.
 */
export type ControlMessage =
    | { kind: 'pointer'; phase: 'move' | 'down' | 'up' | 'cancel'; x: number; y: number; button?: number }
    /** Detents; a fraction is a smooth partial scroll where the desktop supports it. */
    | { kind: 'wheel'; dx: number; dy: number }
    | { kind: 'key'; name?: string; character?: string; down: boolean; modifiers?: string[] }
    | { kind: 'text'; text: string }
    | { kind: 'release_all' }
    | { kind: 'clipboard_read'; request: string }
    | { kind: 'clipboard_write'; request: string; text: string };

export type ControlReply =
    | { kind: 'hello'; protocol: number; geometry: SurfaceGeometry }
    | { kind: 'ack'; seq: number }
    | { kind: 'rejected'; seq: number; code: string; message: string }
    | { kind: 'clipboard'; request: string; text: string; truncated?: boolean; error?: string }
    | { kind: 'revoked'; reason: string };

export function parseControlReply(raw: string): ControlReply | null {
    try {
        const parsed = JSON.parse(raw) as ControlReply;
        return typeof parsed?.kind === 'string' ? parsed : null;
    } catch {
        return null;
    }
}
